glance/common/utils.py

```diff
@@ -65,6 +68,74 @@
 
 GLANCE_TEST_SOCKET_FD_STR = 'GLANCE_TEST_SOCKET_FD'
 
+def fetch_domain_info(project_id, auth_token=None):
+    try:
+        LOG.debug("Starting fetch_domain_info for project_id=%s", project_id)
+
+        # Setup Keystone client
+        if CONF.glance_service_user.enabled:
+            loader = loading.get_plugin_loader("password")
+            auth = loader.load_from_options(
+                auth_url=CONF.keystone_authtoken.auth_url,
+                username=CONF.glance_service_user.username,
+                password=CONF.glance_service_user.password,
+                user_domain_name=CONF.glance_service_user.user_domain_name,
+                project_name=CONF.glance_service_user.project_name,
+                project_domain_name=CONF.glance_service_user.project_domain_name,
+            )
+            sess = session.Session(auth=auth)
+            keystone = keystone_client.Client(session=sess)
+        else:
+            if not auth_token:
+                raise Exception("No auth token provided for user token mode")
+            sess = session.Session(auth=None, token=auth_token)
+            keystone = keystone_client.Client(
+                session=sess, endpoint=CONF.keystone_authtoken.auth_url
+            )
+
+        # Step 1: Get domain_id from project
+        project = keystone.projects.get(project_id)
+        domain_id = getattr(project, "domain_id", None)
+        if not domain_id:
+            LOG.warning("Project %s has no domain_id", project_id)
+            return {"domain_id": "", "domain_name": "", "tags": []}
+
+        LOG.debug("Extracted domain_id=%s from project_id=%s", domain_id, project_id)
+
+        # Step 2: Get domain name
+        domain = keystone.domains.get(domain_id)
+        domain_name = getattr(domain, "name", "")
+        LOG.debug("Extracted domain_name=%s from domain_id=%s", domain_name, domain_id)
+
+        # Step 3: List tags on domain_id (as project)
+        LOG.debug("Calling keystone.projects.list_tags(domain_id=%s)", domain_id)
+        tags_response = keystone.projects.list_tags(domain_id)
```

Comment:
You opted to use `list_tags()` after all instead of `projects.get(domain_id)`. I thought this added the requirement of a `reader` role to the user token. But since you use the glance service user by default, I guess this doesn't matter anymore?

I'd still like to see the reason for that choice documented somewhere, e.g. here or in the commit message.

Given that a normal user token probably doesn't work, do you even need the code-path that uses the user-token?